packages/react/src/jsx.tsx

```diff
@@ -0,0 +1,48 @@
+import { JSXElementConstructor, ReactElement } from 'react'
+
+import { handler } from './events'
+import { ComponentDataState, JSONValue } from './template'
+
+type Props = Record<string, JSONValue | ((args: any) => any)> & {
+  children: JSONValue
+}
+
+// tbd: investigate this
+export const jsxs = jsx
+
+export function jsx(
+  componentFactory: (props: any) => React.ReactElement<Props>,
+  { children, ...passedProps }: Props,
+  key: string
+): ComponentDataState {
+  const { type, props } = componentFactory(passedProps)
+  if (typeof type !== 'string') {
+    throw new Error('Invalid component. Make sure to use server-side version of your components.')
+  }
+  const serialisedProps = Object.fromEntries(
+    Object.entries(props).map(([key, value]) => {
+      if (typeof value === 'function') {
+        return [key, handler(value)]
+      }
+      return [key, value]
+    })
+  )
+  return {
+    $: 'component',
+    component: type,
+    key,
+    props: serialisedProps,
+    children,
+  }
+}
+
+export function createComponentDefinition<
+  T extends JSXElementConstructor<any> | keyof JSX.IntrinsicElements,
+  P = React.ComponentProps<T>,
+>(type: string) {
+  return (props: P): ReactElement => ({
+    type,
+    props,
```

Comment:
Right now, TypeScript does not allow to change the type of JSX factory to something non-standard. To avoid type errors, we have to return "ReactElement". So, `type`, `props`, and `key`. We don't need key (yet), and also, it's computed on the client.

Reply:
Ah. Well, we do need key, right? It may be necessary for some devs to re-order the elements of an array, and key is the only way

Reply:
Yes - correct. If you look at `jsx` function factory, we get `key` and pass it to the client, so you can set keys. This is just a "componentDefinition" function that allows to write <Component /> on the server. 

```ts
export function jsx(
  componentFactory,
  { children, ...passedProps },
  key,
): ComponentDataState {
  const { type, props } = componentFactory(passedProps)
  return {
      $: 'component',
      component: type,
      key,
      props,
      children,
   }
}
```  

So, here's few things happening:
- `componentFactory` takes `props` and returns `props`. This is not really necessary - because we use `componentFactory` really to just get `type` of component (its stringified name). I left it this way in case we wanted to add validation on the server-side in the future. If yes, that would happen inside `componentFactory`
- we do not pass `key` to component factory, because its purpose is to just get component type and (in the future) props - as per previous point. We still put key on the "ComponentDataState"

Reply:
The last bullet is the only one that I don't understand. What is the `componentDefinition` in comparison to the `componentFactory`?

Reply:
```
componentFactory: ReturnType<typeof createComponentDefinition>,
```